MIDAS/src/finite-state-machines/fsm.cpp

```diff
@@ -0,0 +1,367 @@
+
+        case STATE_MAIN_DEPLOY:
+            // if detected a sharp change in jerk then go to the next state
+            if(getJerk(hg) < 0) { 
+		        state.curr_state = STATE_MAIN;
+                break;
+            }
+
+            // if no transtion after a certain amount of time then just move on to next state
+            if(current_time - main_time > main_to_main_deploy_timer_threshold) {
+                state.curr_state = STATE_MAIN;
+            }
+            break;
+
+        case STATE_MAIN:
+            // if slowed down enough then go on to the next state
+            if (abs(getVerticalSpeed(bar)) <= landed_vertical_speed_threshold) { 
+             	landed_time = current_time;
+             	state.curr_state = STATE_LANDED; 
+            }
+            break;
+
+        case STATE_LANDED:
+            // if the slow speed was too brief then return to previous state
+            if (abs(getVerticalSpeed(bar)) > landed_vertical_speed_threshold && current_time - landed_time > landed_timer_threshold) {
+             	state.curr_state = STATE_MAIN; 
+            }
+            break;
+        
+        default:
+            return state;
+    
+    }
+    return state;
+}
+
+// this is similiar to the previous function but contains less states
+FSMState FSM::tick_fsm_booster(FSMState state, std::array<HighGData, 8> hg, std::array<Barometer, 8> bar){
+
+    double current_time = tick_to_ms(xTaskGetTickCount());
+    
+    switch (state.curr_state)
+    {
+        case STATE_IDLE:
+            if(getAcceleration(hg) > idle_to_first_boost_acceleration_threshold) {
+                launch_time = current_time;
+                state.curr_state = STATE_FIRST_BOOST;
+            }
+            
+            break;
+
+        case STATE_FIRST_BOOST:
+            if(getAcceleration(hg) < idle_to_first_boost_acceleration_threshold && current_time - launch_time < idle_to_first_boost_time_threshold) {
+                 state.curr_state = STATE_IDLE;
+                 break;
+            }
+            if(getAcceleration(hg) < coast_detection_acceleration_threshold) {
+                burnout_time = current_time;
+                state.curr_state  = STATE_BURNOUT;
+            }
+
+            break;
+
+        case STATE_BURNOUT:
+            if(getAcceleration(hg) >= coast_detection_acceleration_threshold && (current_time - burnout_time) < first_boost_to_burnout_time_threshold) { 
+                state.curr_state = STATE_FIRST_BOOST;
+                break;
+            }
+
+            if(current_time - burnout_time > first_boost_to_burnout_time_threshold) { 
+                sustainer_ignition_time = current_time;
+                state.curr_state = STATE_SUSTAINER_IGNITION;
+            }
+            break;
+
+        case STATE_FIRST_SEPERATION: 
+            if (getJerk(hg) < 0) {
+                state.curr_state = STATE_COAST;
+                break;
+            }
+
+            first_seperation_time = current_time;
```

Comment:
this value gets updated to be current time right before checking the difference between itself and current time, so the next if statement will always evaluate to false. We want to set this value when we transition to STATE_FIRST_SEPARATION, not while we are in it